File: leetcode/problems/python3/3700. Number of ZigZag Arrays II/solve.py
class Solution:
    def zigZagArrays(self, n: int, l: int, r: int) -> int:
        MOD = 10**9 + 7

        matrix_size = r - l

        if matrix_size == 1:
            return 2

        trans_matrix = [[0] * matrix_size for _ in range(matrix_size)]

        for row in range(matrix_size):
            for col in range(matrix_size - row - 1, matrix_size):
                trans_matrix[row][col] = 1

        init_vector = [i + 1 for i in range(matrix_size)]

        def mat_mul(mat_a, mat_b):
            size = len(mat_a)
            res = [[0] * size for _ in range(size)]
            for i in range(size):
                r_i = res[i]
                a_i = mat_a[i]
                for k in range(size):
                    a_ik = a_i[k]
                    if a_ik:
                        b_k = mat_b[k]
                        for j in range(size):
                            r_i[j] = (r_i[j] + a_ik * b_k[j]) % MOD
            return res

        def mat_vec_mul(mat, vector):
            size = len(mat)
            res = [0] * size
            for i in range(size):
                a_i = mat[i]
                val = 0
                for j in range(size):
                    val = (val + a_i[j] * vector[j]) % MOD
                res[i] = val
            return res

        def mat_pow(mat, power):
            size = len(mat)
            res = [[0] * size for _ in range(size)]
            for i in range(size):
                res[i][i] = 1
            base = mat
            while power > 0:
                if power % 2 == 1:
                    res = mat_mul(res, base)
                base = mat_mul(base, base)
                power //= 2
            return res

        exp_times = n - 2

        trans_matrix = mat_pow(trans_matrix, exp_times)

        result_vector = mat_vec_mul(trans_matrix, init_vector)

        return ((sum(result_vector) % MOD) * 2) % MOD

File: leetcode/problems/go/3700. Number of ZigZag Arrays II/solve.go
package main

const MOD = 1_000_000_007

func matMul(matA, matB [][]int) [][]int {
	size := len(matA)
	res := make([][]int, size)
	for i := range res {
		res[i] = make([]int, size)
	}

	for i := range size {
		aI := matA[i]
		for k := range size {
			aIK := aI[k]
			if aIK == 0 {
				continue
			}
			bK := matB[k]
			rI := res[i]
			for j := range size {
				rI[j] = (rI[j] + aIK*bK[j]) % MOD
			}
		}
	}
	return res
}

func matPow(mat [][]int, exp int) [][]int {
	size := len(mat)
	res := make([][]int, size)
	for i := range res {
		res[i] = make([]int, size)
		res[i][i] = 1
	}

	base := mat
	for exp > 0 {
		if exp%2 == 1 {
			res = matMul(res, base)
		}
		base = matMul(base, base)
		exp /= 2
	}
	return res
}

func matVecMul(mat [][]int, vec []int) []int {
	size := len(mat)
	res := make([]int, size)
	for i := range size {
		aI := mat[i]
		sum := 0
		for j := range size {
			sum = (sum + aI[j]*vec[j]) % MOD
		}
		res[i] = sum
	}
	return res
}

func zigZagArrays(n int, l int, r int) int {
	K := r - l

	if K == 1 {
		return 2
	}

	transMatrix := make([][]int, K)
	for i := range transMatrix {
		transMatrix[i] = make([]int, K)
	}

	for row := range K {
		for col := K - row - 1; col < K; col++ {
			transMatrix[row][col] = 1
		}
	}

	initVec := make([]int, K)
	for i := range K {
		initVec[i] = i + 1
	}

	expTimes := n - 2
	finalTransMatrix := matPow(transMatrix, expTimes)
	resultVec := matVecMul(finalTransMatrix, initVec)
	totalSum := 0
	for _, v := range resultVec {
		totalSum = (totalSum + v) % MOD
	}

	return (totalSum * 2) % MOD
}
